lib/sshutils/sftp/sftp.go

```diff
@@ -323,94 +320,112 @@ func (c *Config) initFS(sshClient *ssh.Client, client *sftp.Client) error {
 	}
 
 	if c.getHomeDir == nil {
-		c.getHomeDir = func() (string, error) {
-			return getRemoteHomeDir(sshClient)
-		}
+		c.getHomeDir = getLocalHomeDir
 	}
 
 	return trace.Wrap(c.expandPaths(srcOK, dstOK))
 }
 
 func (c *Config) expandPaths(srcIsRemote, dstIsRemote bool) (err error) {
-	srcHomeRetriever := getLocalHomeDir
-	if srcIsRemote {
-		srcHomeRetriever = c.getHomeDir
-	}
 	for i, srcPath := range c.srcPaths {
-		c.srcPaths[i], err = expandPath(srcPath, srcHomeRetriever)
+		c.srcPaths[i], err = expandPath(srcPath, srcIsRemote, c.getHomeDir)
 		if err != nil {
-			return trace.Wrap(err)
+			return trace.Wrap(err, "error expanding %q", srcPath)
 		}
 	}
 
-	dstHomeRetriever := getLocalHomeDir
-	if dstIsRemote {
-		dstHomeRetriever = c.getHomeDir
-	}
-	c.dstPath, err = expandPath(c.dstPath, dstHomeRetriever)
-
-	return trace.Wrap(err)
+	c.dstPath, err = expandPath(c.dstPath, dstIsRemote, c.getHomeDir)
+	return trace.Wrap(err, "error expanding %q", c.dstPath)
 }
 
-func getLocalHomeDir() (string, error) {
-	u, err := user.Current()
-	if err != nil {
-		return "", trace.Wrap(err)
+func expandPath(pathStr string, isRemotePath bool, getHomeDir homeDirRetriever) (string, error) {
+	pfxLen := homeDirPrefixLen(pathStr)
+	if pfxLen == -1 {
+		return pathStr, nil
 	}
-	return u.HomeDir, nil
-}
 
-func expandPath(pathStr string, getHomeDir homeDirRetriever) (string, error) {
-	if !needsExpansion(pathStr) {
-		return pathStr, nil
+	// if the path is local, expand the path to be rooted at this user's
+	// home dir
+	if !isRemotePath {
+		var username string
+		// if the path starts with ~user, resolve that user's home dir
+		// otherwise resolve the current user's home dir
+		if pfxLen == 1 && len(pathStr) > 1 {
+			name, _, ok := strings.Cut(pathStr, "/")
+			// if we didn't find anything with back slashes, try with
+			// forward slashes when on Windows
+			if runtime.GOOS == "windows" && !ok {
+				name, _, ok = strings.Cut(pathStr, `\`)
+			}
+			if ok {
+				// skip the initial '~'
+				pfxLen = len(name)
+				username = name[1:]
+			}
+		}
+
+		homeDir, err := getHomeDir(username)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		// this is safe because we verified that all paths are non-empty
+		// in CreateUploadConfig/CreateDownloadConfig
+		return path.Join(homeDir, pathStr[pfxLen:]), nil
```

Comment:
I'd use `filepath.Join` instead, it will be another situation where you don't have to think about OS specific separators.

Reply:
SFTP requires paths separated by `/`, so the thought was to just use `path.Join` here so that's enforced. It doesn't really matter since paths are `path.Clean`ed anyway. Regardless, this code is now removed given your comment below.